storage/jellyfish-merkle/src/lib.rs

```diff
@@ -389,102 +419,125 @@ where
         kvs: &[(HashValue, &(HashValue, K))],
         depth: usize,
         hash_cache: &Option<&HashMap<NibblePath, HashValue>>,
-        tree_cache: &mut TreeCache<R, K>,
+        batch: &mut TreeUpdateBatch<K>,
     ) -> Result<(NodeKey, Node<K>)> {
-        let node = tree_cache.get_node(&node_key)?;
+        let node = self.reader.get_node(&node_key)?;
+        batch.put_stale_node(node_key.clone(), version);
+
         Ok(match node {
             Node::Internal(internal_node) => {
-                // We always delete the existing internal node here because it will not be referenced anyway
-                // since this version.
-                tree_cache.delete_node(&node_key, false /* is_leaf */);
-
                 // Reuse the current `InternalNode` in memory to create a new internal node.
                 let mut children: Children = internal_node.clone().into();
 
                 // Traverse all the path touched by `kvs` from this internal node.
-                for (left, right) in NibbleRangeIterator::new(kvs, depth) {
-                    // Traverse downwards from this internal node recursively by splitting the updates into
-                    // each child index
-                    let child_index = kvs[left].0.get_nibble(depth);
-
-                    let (new_child_node_key, new_child_node) =
-                        match internal_node.child(child_index) {
-                            Some(child) => {
-                                let child_node_key =
-                                    node_key.gen_child_node_key(child.version, child_index);
-                                self.batch_insert_at(
-                                    child_node_key,
-                                    version,
-                                    &kvs[left..=right],
-                                    depth + 1,
-                                    hash_cache,
-                                    tree_cache,
-                                )?
-                            }
-                            None => {
-                                let new_child_node_key =
-                                    node_key.gen_child_node_key(version, child_index);
-                                self.batch_create_subtree(
-                                    new_child_node_key,
+                let range_iter = NibbleRangeIterator::new(kvs, depth);
+                let new_children: Vec<_> = if depth <= MAX_PARALLELIZABLE_DEPTH {
```

Comment:
why do you want to have this MAX_PARALLELIZABLE_DEPTH rather than blindly put everything into the thread pool?

Reply:
To reduce the cost of inter thread communication. It does make things faster as I benchmarked the SMT back early last year.

btw `.with_min_len(100)` that you see elsewhere in the repo serves the same purpose.

Reply:
Sorry I didn't fully understand. If we want to minimize the cost, why not swap the if and else sections so that a subtree after a certain level will be on the same thread (or did I misread the code?)

Reply:
That was the intention.... goooooood catch..... :trollface: :trollface: :trollface: